storagenode/pieces/cache.go

```diff
@@ -0,0 +1,163 @@
+func (store *StoreWithCache) SpaceUsedForPiecesLive(ctx context.Context) int64 {
+	store.cacheMutex.Lock()
+	defer store.cacheMutex.Unlock()
+	return store.cache.total
+}
+
+// SpaceUsedBySatelliteLive returns the current total space used for a specific
+// satellite for all pieces (not including header bytes)
+func (store *StoreWithCache) SpaceUsedBySatelliteLive(ctx context.Context, satelliteID storj.NodeID) int64 {
+	store.cacheMutex.Lock()
+	defer store.cacheMutex.Unlock()
+	return store.cache.totalBySatellite[satelliteID]
+}
+
+// UpdateCache updates the live used space totals
+// with a pieceSize that was either created or deleted where the pieceSize is
+// only the content size and does not include header bytes
+func (store *StoreWithCache) UpdateCache(ctx context.Context, satelliteID storj.NodeID, pieceSize int64) {
+	store.cacheMutex.Lock()
+	defer store.cacheMutex.Unlock()
+	store.cache.total += pieceSize
+	store.cache.totalBySatellite[satelliteID] += pieceSize
+}
+
+// RecalculateSpaceUsedCache sums up the bytes for all pieces (not including headers) currently stored.
+// The live values for used space are initially calculated when the storagenode starts up, then
+// incrememted/decremeted when pieces are created/deleted. In addition we want to occasionally recalculate
+// the live values to confirm correctness. This method RecalculateSpaceUsedCache is responsible for doing that.
+func (store *StoreWithCache) RecalculateSpaceUsedCache(ctx context.Context) error {
+	store.cacheMutex.Lock()
+	// Save the current live values before we start recalculating
+	// so we can compare them to what we recalculate
+	spaceUsedWhenIterationStarted := store.cache
+	store.cacheMutex.Unlock()
+
+	spaceUsedResultOfIteration, err := store.spaceUsedForPiecesAndBySatellitesSlow()
+	if err != nil {
+		return err
+	}
+
+	store.cacheMutex.Lock()
+	defer store.cacheMutex.Unlock()
+	// Since it might have taken a long time to iterate over all the pieces, here we need to check if
+	// we missed any writes/deletes of pieces while we were iterating and estimate
+	store.estimateAndSave(spaceUsedWhenIterationStarted, spaceUsedResultOfIteration)
+
+	return nil
+}
+
+func (store *StoreWithCache) estimateAndSave(oldSpaceUsed, newSpaceUsed spaceUsedCache) {
+	estimatedTotalsBySatellites := map[storj.NodeID]int64{}
+	for satelliteID, newSpaceUsedTotal := range newSpaceUsed.totalBySatellite {
+		estimatedTotalsBySatellites[satelliteID] = estimateRecalculation(newSpaceUsedTotal,
+			oldSpaceUsed.totalBySatellite[satelliteID],
+			store.cache.totalBySatellite[satelliteID],
+		)
+	}
+
+	store.cache.totalBySatellite = estimatedTotalsBySatellites
+	store.cache.total = estimateRecalculation(newSpaceUsed.total,
+		oldSpaceUsed.total,
+		store.cache.total,
+	)
+}
+
+func estimateRecalculation(newSpaceUsedTotal, spaceUsedWhenIterationStarted, spaceUsedWhenIterationEnded int64) int64 {
+	if newSpaceUsedTotal == spaceUsedWhenIterationEnded {
```

Comment:
This shortcut might actually be wrong. Check me on this:

1. spaceUsedWhenIterationStarted = 10
2. meanwhile on disk there's actually 30 bytes being used. 20 of bytes just didn't get counted for whatever reason
3. while iterating over blob directories, 4 blobs get written at 10 bytes each
4. half of those blobs are noticed by the directory iteration. it returns 50 (30 + 2 * 10)
5. spaceUsedWhenIterationEnded = 50 (10 + 4 * 10)
6. estimateRecalculation sees 50 == 50, shortcuts, returns 50
7. but the right answer would have been 70 (30 + 4 * 10), which we would have gotten if we had followed through the half-delta adding below (50 + (50 - 10) / 2)